src/game.py

```diff
@@ -0,0 +1,207 @@
+import pygame
+from pygame.time import Clock
+from threading import Semaphore
+from typing import Optional
+from threading import Thread
+
+from src.client.game_client import Client
+from src.map.game_map import Map
+from src.players.player_factory import PlayerFactory
+from src.players.player import Player
+
+
+class Game(Thread):
+    def __init__(self, name: str = None, max_players: int = 1, num_turns: int = None, is_full: bool = False) -> None:
+        super().__init__()
+
+        self.__name: str = name
+        self.map: Optional[Map] = None
+        self.running: bool = True
+        self.__winner: Optional[int] = None
+        self.__is_full: bool = is_full
+
+        self.num_turns: int = num_turns
+        self.current_turn: Optional[int] = None
+        self.num_rounds: Optional[int] = None
+        self.current_round: Optional[int] = None
+        self.__round_started: bool = False
+
+        self.__current_client: Client = Client()
+        self.__all_clients: dict[Player, Client] = {}
+
+        self.__current_player: Optional[Player] = None
+        self.__waiting_players: list[Player] = []
+        self.__players_in_game: dict[int, Player] = {}
+        self.__player_wins: dict[int, int] = {}
+
+        self.__current_player_idx: Optional[int] = None
+        self.__max_players: int = max_players
+        self.__game_players: int = 0
+        self.__all_players: int = 0
+
+        self.__turn_played_sem: Semaphore = Semaphore(0)
+        self.__clock: Clock = Clock()
+
+    def add_player(self, name: str, password: str = None, is_observer: bool = None) -> None:
+        if self.__game_players >= self.__max_players:
+            is_observer = True
+
+        if not is_observer:
+            self.__game_players += 1
+
+        self.__all_players += 1
+
+        player: Player
+        player_type: str
+
+        if is_observer:
+            player_type = "observer"
+        else:
+            player_type = "bot_player"
+
+        player = PlayerFactory.create_player(player_type, name, self.__turn_played_sem, self.__current_player_idx,
+                                             self.__game_players - 1, self.running, password, is_observer)
+
+        self.__waiting_players.append(player)
+
+    def run(self) -> None:
+        self.__start_game()
+
+        while self.running:
+
+            if not self.__round_started:
+                self.__update_round()
+
+            self.__update_turn()
+
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:
+                    self.running = False
+
+            self.map.draw_map(self.current_turn, self.num_turns, self.current_round, self.num_rounds)
+
+            self.__clock.tick(60)
+
+            for player in self.__players_in_game.values():
+                player.next_turn_sem.release()
```

Comment:
I don't really see the point of using semaphores here, every tick only 1 bot needs to make a full turn, and the action of the turn for the rest is not expensive